Change diag(A::StructuredMatrix[, k=0])

fredrikekre opened the pull request:
My take on JuliaLang/LinearAlgebra.jl#478 
Changes:
 - `diag(A::StructuredMatrix[, k=0])` now return a new vector, such that it does not alias the wrapped vector. This now behaves the same for eg `Matrix`, where we get a new vector. It is also consistent with `getindex` (`diag` is essentially a special case of `getindex`).
 - `diag(A::StructuredMatrix{T,VectorType}[, k=0])` always return a vector of type `VectorType`, i.e. the same type which the diagonals are stored as (actually we return `similar(::VectorType, k::Int)`, but that is usually of type `VectorType`). ~~This is problematic for wrapped `UnitRange`s but that is already pretty broken for this case.~~ EDIT: The latest revision does actually work for `UnitRange`s also, since `similar(1:4)` return an `Vector{Int}` :tada: 

fix JuliaLang/LinearAlgebra.jl#478

fredrikekre commented on base/linalg/bidiag.jl:
```diff
@@ -272,13 +272,13 @@ end
-        return M.uplo == 'U' ? M.ev : zeros(T, size(M,1)-1)
-    elseif n == -1
-        return M.uplo == 'L' ? M.ev : zeros(T, size(M,1)-1)
+        return copy(M.dv)
```
Should change this to `copy!(similar(M.dv, length(M.dv)), M.dv)` to make certain we return the same type, since we then call the same `similar` method no matter what `n` is.

Sacha0 commented on base/linalg/bidiag.jl:
```diff
@@ -270,15 +270,13 @@ function triu!(M::Bidiagonal, k::Integer=0)
-        return M.uplo == 'U' ? M.ev : zeros(T, size(M,1)-1)
-    elseif n == -1
-        return M.uplo == 'L' ? M.ev : zeros(T, size(M,1)-1)
+        return copy!(similar(M.dv, length(M.dv)), M.dv)
```
Could the `length(M.dv)` be omitted?

Sacha0 replied:
Moreover, why not simply `copy(...` rather than `copy!(similar(M.dv, length(M.dv)), M.dv)`? Is the return type dependent upon `k` in some cases (e.g. for ranges) otherwise?

fredrikekre replied:
> Guarding against the length specification modifying the return type?

Yes, that was the idea, to guarantee that we take the same path no matter which diagonal was requested.

> Moreover, why not simply copy(... rather than copy!(similar(M.dv, length(M.dv)), M.dv)? Does the return type otherwise depend upon k in some cases (e.g. for ranges)?

Same comment applies; guarantee the same code path. Also `copy(r::AbstractRange) = r` so that will not yield a `Vector{Int}` as `similar` does.
```julia
julia> copy(1:4)
1:4
```

I don't think the `copy!(similar(...), ...)` uses are that terrible -- it is pretty clear what is going on IMO.

Sacha0 replied:
Sounds good then! I might consider adding comments to make the intention clear for future readers; to someone not acquainted with the subtleties involved, the intention may be far from clear :).

fredrikekre replied:
Yea, Ill add that

Sacha0 reviewed (approved):
If my guesses above at the reasons for the particular `copy!(similar(...` invocations are correct, then lgtm! :)

fredrikekre committed:
rewrite diag(::StructuredMatrix[, k=0])

 - diag(A::StructuredMatrix[, k=0]) now return a new vector,
   such that it does not alias the wrapped vector. This now
   behaves the same for eg Matrix, where we get a new vector.
   It is also consistent with getindex (diag is essentially a
   special case of getindex).

 - diag(A::StructuredMatrix{T,VectorType}[, k=0]) now call
   similar(..., ::Int) on every branch to make sure the same
   vector type is always returned, this is usually of type
   VectorType.

Sacha0 commented on base/linalg/bidiag.jl:
```diff
@@ -270,15 +270,15 @@ function triu!(M::Bidiagonal, k::Integer=0)
-function diag(M::Bidiagonal{T}, n::Integer=0) where T
+function diag(M::Bidiagonal, n::Integer=0)
+    # every branch call similar(..., ::Int) to make sure the
+    # same vector type is returned independent of n
```
:+1:  :)

Sacha0 reviewed (approved):
Thanks @fredrikekre! :)
